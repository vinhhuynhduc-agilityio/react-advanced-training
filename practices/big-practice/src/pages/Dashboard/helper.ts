import { GridApi } from 'ag-grid-community';

/**
 * Handles scrolling to new user or task
 * Use setTimeOut to ensure `getRowNode` is called
 * after the table has re-rendered and refreshed the data
 */
const handleScrollingToAddedRow = (
  id: string,
  gridApi: GridApi
) => {
  setTimeout(() => {
    if (gridApi) {
      const rowNode = gridApi.getRowNode(id);

      if (rowNode) {
        gridApi.ensureNodeVisible(rowNode, 'middle');
      }
    }
  }, 100);
};

const handleRowSelection = (
  userId: string | null,
  sourceComponent: string,
  handleRowSelected: (userId: string | null, sourceComponent: string) => void
) => {
  handleRowSelected(userId, sourceComponent);
};

export {
  handleScrollingToAddedRow,
  handleRowSelection
};