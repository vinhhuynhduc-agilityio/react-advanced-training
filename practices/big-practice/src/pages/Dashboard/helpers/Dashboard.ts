import { GridApi } from "ag-grid-community";

// Returns the current date and time as a formatted string in the format "Nov 10, 2024 20:54:10".
const getRegisteredDate = () =>
  `${new Date().toLocaleDateString("en-US", {
    month: "short",
    day: "2-digit",
    year: "numeric",
  })} ${new Date().toLocaleTimeString("en-US", {
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
  })}`;

//  Format a given Date object into a string with the format "24 Jan 23".
const formatStartDate = (date: Date): string => {
  return date.toLocaleDateString("en-GB", {
    day: "2-digit",
    month: "short",
    year: "2-digit",
  });
};

/**
 * Handles scrolling to new user or task
 * Use setTimeOut to ensure `getRowNode` is called
 * after the table has re-rendered and refreshed the data
 */
const handlesScrollingToNewUserOrTask = (
  id: string,
  gridApi: GridApi
) => {
  setTimeout(() => {
    if (gridApi) {
      const rowNode = gridApi.getRowNode(id);

      if (rowNode) {
        gridApi.ensureNodeVisible(rowNode, 'middle');
      }
    }
  }, 100);
};

export {
  getRegisteredDate,
  formatStartDate,
  handlesScrollingToNewUserOrTask
};
