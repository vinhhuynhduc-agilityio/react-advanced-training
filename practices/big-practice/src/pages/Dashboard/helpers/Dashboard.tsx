// Returns the current date and time as a formatted string in the format "Nov 10, 2024 20:54:10".
const getRegisteredDate = () => `${new Date().toLocaleDateString('en-US', {
  month: 'short',
  day: '2-digit',
  year: 'numeric',
})} ${new Date().toLocaleTimeString('en-US', {
  hour: '2-digit',
  minute: '2-digit',
  second: '2-digit',
  hour12: false,
})}`;

export {
  getRegisteredDate
};
