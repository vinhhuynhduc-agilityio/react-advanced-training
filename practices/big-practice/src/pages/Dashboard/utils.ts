export const handleRowSelection = (
  userId: string | null,
  sourceComponent: string,
  handleRowSelected: (userId: string | null, sourceComponent: string) => void
) => {
  handleRowSelected(userId, sourceComponent);
};
