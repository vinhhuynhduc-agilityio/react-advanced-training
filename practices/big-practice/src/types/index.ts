export * from './chart';
export * from './task';
export * from './user';
export * from './project';
