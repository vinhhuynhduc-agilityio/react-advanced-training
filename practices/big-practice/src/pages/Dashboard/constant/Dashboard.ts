import { UserData } from "@/types/table";

const initialDefaultValues: UserData = {
  fullName: "",
  email: "",
  avatarUrl: "https://via.placeholder.com/80",
  id: "",
  earnings: "",
  registered: "",
  lastUpdated: ""
};

export {
  initialDefaultValues
};
