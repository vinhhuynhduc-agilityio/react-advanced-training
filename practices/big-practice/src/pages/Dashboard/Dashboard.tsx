import React,
{
  useCallback,
  useEffect,
  useRef,
  useState
} from "react";
import { v4 as uuidv4 } from 'uuid';

// components
import Header from "@/components/Header/Header";
import UserListDrawer from "./UserListDrawer/UserListDrawer";
import TaskDashboard from "./TaskDashboard/TaskDashboard";
import Footer from "@/components/Footer/Footer";
import ModalDialog from "@/components/ModalDialog/ModalDialog";
import UserProfileForm from "./UserProfileForm/UserProfileForm";
import ProjectForm from "./ProjectForm/ProjectForm";
import TaskForm from "./TaskForm/TaskForm";
import ChartTotalTasksCompleted from "./ChartTotalTasksCompleted/ChartTotalTasksCompleted";
import ChartTotalTasksByProjects from "./ChartTotalTasksByProjects/ChartTotalTasksByProjects";
import ChartIndividualEmployeeProgress from "./ChartIndividualEmployeeProgress/ChartIndividualEmployeeProgress";

// types
import {
  UserFormData,
  ProjectsData,
  TaskData,
  UserData
} from "@/types/table";
import { TaskFormData } from "@/types/taskForm";

// utils
import { apiRequest } from "@/services/apiRequest";

// constant
import { initialDefaultValues } from "@/constant/dashboard";

// ag-grid
import { GridApi } from "ag-grid-community";

// helpers
import {
  formatStartDate,
  getRegisteredDate,
  handlesScrollingToNewUserOrTask
} from "../../helpers/dashboard";
import { API_ROUTES } from "@/constant/api";

import { API_BASE_URL } from "@/config";

const Dashboard: React.FC = () => {
  const userListGridApi = useRef<GridApi | null>(null);
  const taskDashboardGridApi = useRef<GridApi | null>(null);

  const [users, setUsers] = useState<UserData[]>([]);
  const [tasks, setTasks] = useState<TaskData[]>([]);
  const [projects, setProjects] = useState<ProjectsData[]>([]);
  const [selectedUserId, setSelectedUser] = useState<string | null>(null);
  const [sourceComponent, setSourceComponent] = useState<string | null>(null);

  // State variables related to opening modal dialog
  const [isModalOpen, setModalOpen] = useState(false);
  const [isEditUser, setEditUser] = useState(false);
  const [defaultValues, setDefaultValues] = useState<UserData>(initialDefaultValues);
  const [isProjectModalOpen, setProjectModalOpen] = useState(false);
  const [isTaskModalOpen, setTaskModalOpen] = useState(false);

  // State to track loading
  const [isLoading, setLoading] = useState(true);

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      try {
        const [
          usersData,
          tasksData,
          projectsData
        ] = await Promise.all([
          apiRequest<UserData[], UserData[]>('GET', `${API_BASE_URL}${API_ROUTES.USERS}`),
          apiRequest<TaskData[], TaskData[]>('GET', `${API_BASE_URL}${API_ROUTES.TASKS}`),
          apiRequest<ProjectsData[], ProjectsData[]>('GET', `${API_BASE_URL}${API_ROUTES.PROJECTS}`),
        ]);
        setUsers(usersData);
        setTasks(tasksData);
        setProjects(projectsData);
      } catch (error) {
        console.error("Failed to fetch data:", error);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, []);


  const registerGridApiTaskDashboard = (api: GridApi) => {
    taskDashboardGridApi.current = api
  };

  // Function to handle when row in TaskDashboard is selected
  const handleTaskRowSelected = (userId: string | null) => {
    setSelectedUser(userId)
    setSourceComponent("TaskDashboard");
  };

  // Function to handle when row in UserListDrawer is selected
  const handleUserRowSelected = (userId: string | null) => {
    setSourceComponent("UserListDrawer");
    setSelectedUser(userId)
  };

  const updateEarningsForUsers = async (
    oldUserId: string,
    newUserId: string,
    currency: number,
    status: boolean
  ) => {
    handleTaskRowSelected(newUserId);

    if (!userListGridApi.current || !status) return;

    setLoading(true);
    const updates: Promise<UserData>[] = [];

    [oldUserId, newUserId].forEach(userId => {
      const rowNode = userListGridApi.current!.getRowNode(userId);

      if (rowNode) {
        const currentEarnings = parseInt(rowNode.data.earnings.slice(1)) || 0;
        const adjustedEarnings = userId === oldUserId
          ? currentEarnings - currency
          : currentEarnings + currency;

        rowNode.setData({
          ...rowNode.data,
          earnings: `$${adjustedEarnings}`
        });

        updates.push(
          apiRequest<UserData, UserData>("PUT", `${API_BASE_URL}${API_ROUTES.USERS}/${userId}`, {
            ...rowNode.data,
            earnings: `$${adjustedEarnings}`,
          })
        );
      }
    });

    try {
      await Promise.all(updates);
    } catch (error) {
      console.error("Failed to update earnings:", error);
    } finally {
      setLoading(false);
    }
  };

  const updateEarningsOnStatusChange = async (
    userId: string,
    currency: number,
    status: boolean
  ) => {
    if (!userListGridApi.current) return;
    setLoading(true);

    // Retrieve `rowNode` by `userId` to update the specific row
    const rowNode = userListGridApi.current.getRowNode(userId);

    if (rowNode) {
      const currentEarnings = parseInt(rowNode.data.earnings.slice(1)) || 0;
      const adjustedEarnings = status
        ? currentEarnings + currency
        : currentEarnings - currency;

      // Update row edit data
      rowNode.setData({
        ...rowNode.data,
        earnings: `$${adjustedEarnings}`
      });

      // Update earnings to backend
      try {
        await apiRequest<UserData, UserData>(
          "PUT",
          `${API_BASE_URL}${API_ROUTES.USERS}/${userId}`,
          {
            ...rowNode.data,
            earnings: `$${adjustedEarnings}`,
          }
        );
      } catch (error) {
        console.error("Failed to update earnings:", error);
      } finally {
        setLoading(false);
      }
    };
  };
  const registerGridApi = (api: GridApi) => {
    userListGridApi.current = api
  };

  const handleToggleProjectForm = useCallback(() => {
    setProjectModalOpen(true);
  }, []);

  const handleToggleTaskForm = useCallback(() => {
    setTaskModalOpen(true);
  }, []);

  const handleToggleUserProfileForm = useCallback(() => {
    setDefaultValues(initialDefaultValues);
    setModalOpen(true);
    setEditUser(false);
  }, []);

  const handleUserDoubleClicked = (userData: UserData) => {
    setDefaultValues(userData);
    setModalOpen(true);
    setEditUser(true);
  };

  const handleOnSubmit = (data: UserFormData) =>
    isEditUser
      ? handleEditUser(data)
      : handleAddNewUser(data);

  const handleAddNewUser = async (data: UserFormData) => {
    const avatarUrl = data.avatarUrl ?? 'https://via.placeholder.com/80';
    const registeredDate = getRegisteredDate();

    setLoading(true);
    const newUser = {
      id: uuidv4(),
      fullName: data.fullName,
      earnings: "$0",
      email: data.email,
      avatarUrl: avatarUrl,
      registered: registeredDate,
      lastUpdated: registeredDate,
    };

    try {
      const addedUser = await apiRequest<UserData, UserData>(
        'POST',
        `${API_BASE_URL}${API_ROUTES.USERS}`,
        newUser
      );

      // Update state users to add new users
      setUsers((prevUsers) => [...prevUsers, addedUser]);

      setModalOpen(false);

      // Handles scrolling to new user
      if (userListGridApi.current) {
        handlesScrollingToNewUserOrTask(newUser.id, userListGridApi.current)
      }

      // handle selected for new user
      handleUserRowSelected(newUser.id);
    } catch (error) {
      console.error("Failed to add new user:", error);
    } finally {
      setLoading(false);
    }
  };

  const handleEditUser = async (data: UserFormData) => {
    if (!defaultValues?.id) return;
    setLoading(true);

    const editUser: UserData = {
      id: defaultValues.id,
      fullName: data.fullName,
      earnings: defaultValues.earnings,
      email: data.email,
      avatarUrl: data.avatarUrl || "https://via.placeholder.com/80",
      registered: defaultValues.registered,
      lastUpdated: getRegisteredDate(),
    };

    try {

      // Send user update request
      const updatedUser = await apiRequest<UserData, UserData>(
        "PUT",
        `${API_BASE_URL}${API_ROUTES.USERS}/${defaultValues.id}`,
        editUser
      );

      setUsers((prevUsers) =>
        prevUsers.map((user) =>
          user.id === defaultValues.id
            ? updatedUser
            : user
        )
      );

      setModalOpen(false);
    } catch (error) {
      console.error("Failed to edit user:", error);
    } finally {
      setLoading(false);
    }
  };

  const handleAddNewTask = async (data: TaskFormData) => {
    const {
      currency,
      project,
      taskName,
      user
    } = data;
    const currencyValue = typeof currency === 'string'
      ? parseInt(currency, 10)
      : currency;

    setLoading(true);

    const newTask = {
      id: uuidv4(),
      userId: user.id,
      projectId: project.id,
      taskName: taskName,
      startDate: formatStartDate(new Date()),
      completedDate: 'incomplete',
      currency: currencyValue,
      status: false,
      projectName: project.value,
      fullName: user.value
    };

    try {
      const addedTask = await apiRequest<TaskData, TaskData>(
        "POST",
        `${API_BASE_URL}${API_ROUTES.TASKS}`,
        newTask
      );

      setTasks((prevTasks) => [...prevTasks, addedTask]);
      setTaskModalOpen(false);

      // Handles scrolling to new user
      if (taskDashboardGridApi.current) {
        handlesScrollingToNewUserOrTask(newTask.id, taskDashboardGridApi.current)
      }

      // handle selected for new task
      handleTaskRowSelected(newTask.userId);
    } catch (error) {
      console.error("Failed to add new task:", error);
    } finally {
      setLoading(false);
    }
  };

  const handleAddProject = async (newProjectName: string) => {
    const newProject = {
      id: uuidv4(),
      projectName: newProjectName
    };
    setLoading(true);

    try {
      const addedProject = await apiRequest<ProjectsData, ProjectsData>(
        'POST',
        `${API_BASE_URL}${API_ROUTES.PROJECTS}`,
        newProject
      );

      setProjects((prevProjects) => [...prevProjects, addedProject]);
      setProjectModalOpen(false);
    } catch (error) {
      console.error("Failed to add new project:", error);
    } finally {
      setLoading(false);
    }
  };

  const renderTaskDashboard = () => {
    return (
      <div className="bg-white mt-4 h-96 border-customBorder">
        <TaskDashboard
          tasks={tasks}
          projects={projects}
          users={users}
          selectedUserId={selectedUserId}
          onTaskRowSelected={handleTaskRowSelected}
          sourceComponent={sourceComponent}
          updateEarningsForUsers={updateEarningsForUsers}
          updateEarningsOnStatusChange={updateEarningsOnStatusChange}
          registerGridApiTaskDashboard={registerGridApiTaskDashboard}
          setTasks={setTasks}
          isLoading={isLoading}
        />
      </div>
    )
  };

  const renderUserListDrawer = () => {
    return (
      <div className="flex-grow-0 ml-1 mr-4 my-4 w-64 ag-theme-alpine overflow-auto">
        <UserListDrawer
          users={users}
          selectedUserId={selectedUserId}
          onUserSelected={handleUserRowSelected}
          sourceComponent={sourceComponent}
          registerGridApi={registerGridApi}
          onUserDoubleClicked={handleUserDoubleClicked}
          isLoading={isLoading}
        />
      </div>
    )
  };

  const renderHeader = () => {
    return (
      <Header
        onAddUser={handleToggleUserProfileForm}
        onAddProject={handleToggleProjectForm}
        onAddTask={handleToggleTaskForm}
      />
    )
  };

  const renderProjectForm = () => (
    <ModalDialog
      title="Add Project"
      onClose={() => setProjectModalOpen(false)}
      content={
        <ProjectForm
          projects={projects}
          onClose={() => setProjectModalOpen(false)}
          onSubmit={handleAddProject}
        />
      }
    />
  );

  const renderTaskForm = () => (
    <ModalDialog
      title="Add Task"
      onClose={() => setTaskModalOpen(false)}
      content={
        <TaskForm
          onClose={() => setTaskModalOpen(false)}
          onSubmit={handleAddNewTask}
          tasks={tasks}
          projects={projects}
          users={users}
        />
      }
    />
  );

  const renderUserProfileForm = () => {
    const title = isEditUser ? "Edit User" : "Add User";
    const buttonLabel = isEditUser ? "Save" : "Add";

    return (
      <ModalDialog
        title={title}
        onClose={() => setModalOpen(false)}
        content={
          <UserProfileForm
            defaultValues={defaultValues}
            isEditUser={isEditUser}
            onClose={() => setModalOpen(false)}
            onSubmit={handleOnSubmit}
            users={users}
            buttonLabel={buttonLabel}
          />
        }
      />
    );
  };

  const renderChartAndTaskContent = () => {
    return (
      <div className="flex-grow bg-slate-100 my-4 mr-4 overflow-auto">
        <ChartTotalTasksCompleted
          tasks={tasks}
          isLoading={isLoading}
        />
        <div className="flex flex-row bg-slate-100 mt-4 h-[302px]">
          <ChartIndividualEmployeeProgress
            tasks={tasks}
            users={users}
            selectedUserId={selectedUserId}
            isLoading={isLoading}
          />
          <ChartTotalTasksByProjects
            tasks={tasks}
            projects={projects}
            isLoading={isLoading}
          />
        </div>
        {renderTaskDashboard()}
      </div>
    )
  };

  const renderContent = () => {
    return (
      <div className="flex flex-row flex-grow bg-slate-100 min-h-0">
        {renderUserListDrawer()}
        {renderChartAndTaskContent()}
      </div>
    )
  };

  return (
    <>
      <div className="flex flex-col h-screen w-screen">
        {renderHeader()}
        {renderContent()}
        <Footer
          content='Team Progress App'
        />
      </div>
      {isModalOpen && renderUserProfileForm()}
      {isProjectModalOpen && renderProjectForm()}
      {isTaskModalOpen && renderTaskForm()}
    </>
  );
};

export default Dashboard;
