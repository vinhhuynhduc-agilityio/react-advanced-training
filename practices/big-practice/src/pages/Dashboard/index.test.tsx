import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { DashboardContext } from '@/context';
import Dashboard from '@/pages/Dashboard';
import { mockContextValue } from '@/mocks';
import { UserData } from '@/types';

// Mocking lazy-loaded components
jest.mock('@/components/UserProfileForm', () => ({
  __esModule: true,
  default: () => <div>UserProfileForm Mock</div>,
}));
jest.mock('@/components/TaskForm', () => ({
  __esModule: true,
  default: () => <div>TaskForm Mock</div>,
}));
jest.mock('@/components/ProjectForm', () => ({
  __esModule: true,
  default: () => <div>ProjectForm Mock</div>,
}));

jest.mock('@/components', () => ({
  ChartIndividualEmployeeProgress: () => <div>ChartIndividualEmployeeProgress Mock</div>,
  ChartTotalTasksByProjects: () => <div>ChartTotalTasksByProjects Mock</div>,
  ChartTotalTasksCompleted: () => <div>ChartTotalTasksCompleted Mock</div>,
  ModalDialog: ({ title, onClose, content }: { title: string, onClose: () => void, content: React.ReactNode }) => (
    <div>
      <div>{title}</div>
      <button onClick={onClose}>Close</button>
      {content}
    </div>
  ),
  Footer: () => <div>Footer Mock</div>,
  Header: ({
    onAddUser,
    onAddTask,
    onAddProject,
  }: {
    onAddUser: () => void;
    onAddTask: () => void;
    onAddProject: () => void;
  }) => (
    <div>
      <button onClick={onAddUser}>Add User</button>
      <button onClick={onAddTask}>Add Task</button>
      <button onClick={onAddProject}>Add Project</button>
    </div>
  ),
  UserListDrawer: ({ onUserDoubleClicked }: { onUserDoubleClicked: (user: UserData) => void }) => (
    <div>
      <button onDoubleClick={() => onUserDoubleClicked({ id: '1', fullName: 'David Smith', earnings: '5000', email: 'Davidsmith@example.com', avatarUrl: '', registered: '', lastUpdated: '' })}>
        Double-click to select user
      </button>
    </div>
  ),
  TaskDashboard: ({ onTaskRowSelected }: { onTaskRowSelected: (task: { id: string; name: string }) => void }) => (
    <div>
      <button onClick={() => onTaskRowSelected({ id: '1', name: 'Row of tasks' })}>
        Row of tasks
      </button>
      <p>TaskDashboard Mock</p>
    </div>
  ),
  ErrorBoundary: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
}));
jest.mock('@/components/common', () => ({ Spinner: () => <div>Spinner Mock</div> }));

describe('Dashboard Component', () => {

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <DashboardContext.Provider value={mockContextValue}>
        <Dashboard />
      </DashboardContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('triggers handleTaskRowSelected on task row selection', async () => {
    render(
      <DashboardContext.Provider value={mockContextValue}>
        <Dashboard />
      </DashboardContext.Provider>
    );

    // Find the task button and simulate a click
    const taskButton = screen.getByText('Row of tasks');
    fireEvent.click(taskButton);
  });


  it('triggers onUserDoubleClicked on row double-click', async () => {
    render(
      <DashboardContext.Provider value={mockContextValue}>
        <Dashboard />
      </DashboardContext.Provider>
    );

    // Find the UserListDrawer and simulate a double-click on the row
    const doubleClickButton = screen.getByText('Double-click to select user');

    // Simulate the double-click event
    fireEvent.doubleClick(doubleClickButton);
    await waitFor(() => screen.getByText('UserProfileForm Mock'));
    expect(screen.getByText('UserProfileForm Mock')).toBeInTheDocument();
  });


  it('renders the "Add User" button and triggers the action', async () => {
    render(
      <DashboardContext.Provider value={mockContextValue}>
        <Dashboard />
      </DashboardContext.Provider>
    );
    const addUserButton = screen.getByText('Add User');
    expect(addUserButton).toBeInTheDocument();

    // Simulate button click
    fireEvent.click(addUserButton);

    // Wait for the UserProfileForm component to be loaded
    await waitFor(() => screen.getByText('UserProfileForm Mock'));

    // Assert that the UserProfileForm component is rendered
    expect(screen.getByText('UserProfileForm Mock')).toBeInTheDocument();
  });

  it('renders the "Add Task" button and triggers the action', async () => {
    render(
      <DashboardContext.Provider value={mockContextValue}>
        <Dashboard />
      </DashboardContext.Provider>
    );
    const addTaskButton = screen.getByText('Add Task');
    expect(addTaskButton).toBeInTheDocument();

    // Simulate button click
    fireEvent.click(addTaskButton);

    // Wait for the UserProfileForm component to be loaded
    await waitFor(() => screen.getByText('TaskForm Mock'));

    // Assert that the UserProfileForm component is rendered
    expect(screen.getByText('TaskForm Mock')).toBeInTheDocument();
  });

  it('renders the "Add Project" button and triggers the action', async () => {
    render(
      <DashboardContext.Provider value={mockContextValue}>
        <Dashboard />
      </DashboardContext.Provider>
    );
    const addProjectButton = screen.getByText('Add Project');
    expect(addProjectButton).toBeInTheDocument();

    // Simulate button click
    fireEvent.click(addProjectButton);

    // Wait for the UserProfileForm component to be loaded
    await waitFor(() => screen.getByText('ProjectForm Mock'));

    // Assert that the UserProfileForm component is rendered
    expect(screen.getByText('ProjectForm Mock')).toBeInTheDocument();
  });

  it('renders the charts correctly', () => {
    render(
      <DashboardContext.Provider value={mockContextValue}>
        <Dashboard />
      </DashboardContext.Provider>
    );
    expect(screen.getByText('ChartTotalTasksCompleted Mock')).toBeInTheDocument();
    expect(screen.getByText('ChartIndividualEmployeeProgress Mock')).toBeInTheDocument();
    expect(screen.getByText('ChartTotalTasksByProjects Mock')).toBeInTheDocument();
  });

  it('renders TaskDashboard and other content correctly', () => {
    render(
      <DashboardContext.Provider value={mockContextValue}>
        <Dashboard />
      </DashboardContext.Provider>
    );
    expect(screen.getByText('TaskDashboard Mock')).toBeInTheDocument();
    expect(screen.getByText('Footer Mock')).toBeInTheDocument();
  });
});
