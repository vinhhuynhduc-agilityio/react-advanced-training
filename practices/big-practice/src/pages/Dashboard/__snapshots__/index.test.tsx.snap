// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col h-screen w-screen"
  >
    <div>
      <button>
        Add User
      </button>
      <button>
        Add Task
      </button>
      <button>
        Add Project
      </button>
    </div>
    <div
      class="flex flex-row flex-grow bg-slate-100 min-h-0"
    >
      <div
        class="flex-grow-0 ml-1 mr-4 my-4 w-64 ag-theme-alpine overflow-auto"
      >
        <div>
          <button>
            Double-click to select user
          </button>
        </div>
      </div>
      <div
        class="flex-grow bg-slate-100 my-4 mr-4 overflow-auto"
      >
        <div>
          <div>
            ChartTotalTasksCompleted Mock
          </div>
        </div>
        <div
          class="flex flex-row bg-slate-100 mt-4 h-[302px]"
        >
          <div>
            <div>
              ChartIndividualEmployeeProgress Mock
            </div>
          </div>
          <div>
            <div>
              ChartTotalTasksByProjects Mock
            </div>
          </div>
        </div>
        <div
          class="bg-white mt-4 h-96 border-customBorder"
        >
          <div>
            <button>
              Row of tasks
            </button>
            <p>
              TaskDashboard Mock
            </p>
          </div>
        </div>
      </div>
    </div>
    <div>
      Footer Mock
    </div>
  </div>
</DocumentFragment>
`;
